src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3574,68 +3574,168 @@ internal Binder GetBinder(CSharpSyntaxNode syntaxNode)
+
+                        case (SourcePropertyAccessorSymbol, SourcePropertyAccessorSymbol):
+                            break; // accessor symbols and their diagnostics are handled by processing the associated property
+
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(prev);
+                    }
+                }
+
+                foreach (var symbol in membersBySignature.Values)
+                {
+                    switch (symbol)
+                    {
+                        case SourceOrdinaryMethodSymbol method:
+                            // partial implementations not paired with a definition
+                            if (method.IsPartialImplementation && method.OtherPartOfPartial is null)
                             {
-                                // Avoid mutating the cached dictionary and especially avoid doing this possibly on multiple threads in parallel.
-                                membersByName = new Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>>(membersByName, ReadOnlyMemoryOfCharComparer.Instance);
+                                diagnostics.Add(ErrorCode.ERR_PartialMethodMustHaveLatent, method.GetFirstLocation(), method);
                             }
+                            else if (method is { IsPartialDefinition: true, OtherPartOfPartial: null, HasExplicitAccessModifier: true })
+                            {
+                                diagnostics.Add(ErrorCode.ERR_PartialMethodWithAccessibilityModsMustHaveImplementation, method.GetFirstLocation(), method);
+                            }
+                            break;
 
-                            membersByName[name] = FixPartialMember(membersByName[name], prevPart, methodPart);
-                        }
+                        case SourcePropertySymbol property:
+                            if (property.OtherPartOfPartial is null)
+                            {
+                                diagnostics.Add(
+                                    property.IsPartialDefinition ? ErrorCode.ERR_PartialPropertyMissingImplementation : ErrorCode.ERR_PartialPropertyMissingDefinition,
+                                    property.GetFirstLocation(),
+                                    property);
+                            }
+                            break;
+
+                        case SourcePropertyAccessorSymbol:
+                            break; // handled by SourcePropertySymbol case
+
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(symbol);
                     }
-                    else
+                }
+            }
+
+            memberNames.Free();
+
+            void mergePartialMethods(ref Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>> membersByName, ReadOnlyMemory<char> name, SourceOrdinaryMethodSymbol currentMethod, SourceOrdinaryMethodSymbol prevMethod)
```

Comment:
>mergePartialMethods [](http://example.com/codeflow?start=17&length=19)

Consider passing an explicit parameter for `diagnostics` too. Then `this` is the only remaining capture.
If I recall correctly, there are benefits to only capturing `this` in local functions.

Same for `mergePartialProperties` below.